Address -Wnonnull issues and further improvements

kgizdov opened the pull request:
# This Pull request:

## Changes or fixes:
- `RooFit::HistFactory::ConfigParser`: function `ConfigParser::CreateMeasurementFromDriverNode()`fix null dereferences, improve logic, fix some other edge cases, optimise performance
- `TMVA::DataLoader`: calling `.front()` on an empty container is undefined, fix also null ptr dereference
- minor RooFit fixes for null ptr dereferences

## Checklist:

- [x] tested changes locally
- [x] updated the docs (if necessary)

hageboeck commented on roofit/histfactory/src/ConfigParser.cxx:
```diff
@@ -274,115 +272,98 @@ HistFactory::Measurement ConfigParser::CreateMeasurementFromDriverNode( TXMLNode
-      cxcoutEHF << "Found XML attribute in Measurement with no name"  << std::endl;
+  TXMLAttr* curAttr = nullptr;
+  while ( (/**/curAttr = dynamic_cast<TXMLAttr*>(attribIt())/**/) != nullptr ) {
+    const char * curAttrName = curAttr->GetName();  // curAttr is guaranteed non-null
+    const char * curAttrValue = curAttr->GetValue();
```
How about making those `const std::string` and ...

kgizdov replied:
good idea

hageboeck commented on roofit/histfactory/src/ConfigParser.cxx:
```diff
@@ -274,115 +272,98 @@ HistFactory::Measurement ConfigParser::CreateMeasurementFromDriverNode( TXMLNode
+      // rowTitle=curAttr->GetValue();
+      measurement.SetName(curAttrValue);
+      // measurement.OutputFileName = outputFileNamePrefix + "_" + rowTitle + ".root";
+    } else if ( curAttrName == TString("Lumi") ) {
+      measurement.SetLumi(atof(curAttrValue));
+    } else if ( curAttrName == TString("LumiRelErr") ) {
+      measurement.SetLumiRelErr(atof(curAttrValue));
+    } else if ( curAttrName == TString("BinLow") ) {
+      measurement.SetBinLow(atoi(curAttrValue));
+    } else if ( curAttrName == TString("BinHigh") ) {
```
... and making those
```c++
else if ( curAttrName == "LumiRelErr" )
```
?

There's no need to create new TStrings for every `if`.

kgizdov replied:
True, wasn't sure how you preferred this internally. I'll update.

hageboeck commented on roofit/histfactory/src/ConfigParser.cxx:
```diff
@@ -391,145 +372,114 @@ HistFactory::Measurement ConfigParser::CreateMeasurementFromDriverNode( TXMLNode
-      map<string,double> gammaSyst;
-      map<string,double> uniformSyst;
-      map<string,double> logNormSyst;
+      // these are currently unused for some reason
+      // std::map<string, double> gammaSyst;
+      // std::map<string, double> uniformSyst;
+      // std::map<string, double> logNormSyst;
```
You can remove invalid code.

hageboeck commented on roofit/roofitcore/src/RooAddModel.cxx:
```diff
@@ -105,7 +105,7 @@ RooAddModel::RooAddModel(const char *name, const char *title, const RooArgList&
           << ") number of pdfs and coefficients inconsistent, must have Npdf=Ncoef or Npdf=Ncoef+1";
     const std::string msgStr = msgSs.str();
-    coutE(InputArguments) << msgStr << endl;
+    coutE(InputArguments) << msgStr << "\n";
```
I wonder if all systems flush the stream (i.e. print to the terminal or the respective message handler) before the exception arrives?
Maybe it's a good idea to let it flush?

kgizdov replied:
I believe this is already handled [here](https://github.com/root-project/root/blob/5aee32a4c137a6ed31ee199ee8c80a6291f2d2e6/roofit/roofitcore/src/RooMsgService.cxx#L500-L506)
and in reality I don't think even `std::endl` can guarantee this. This is OS-dependent and possibly only can be done with [`std::set_terminate`](https://en.cppreference.com/w/cpp/error/set_terminate).

hageboeck replied:
I don't think it's important, so no need to insist. 👍

hageboeck commented on roofit/roofitcore/src/RooRealSumFunc.cxx:
```diff
@@ -152,8 +152,9 @@ RooRealSumFunc::RooRealSumFunc(const char *name, const char *title, const RooArg
    if (func) {
       if (!dynamic_cast<RooAbsReal *>(func)) {
-         coutE(InputArguments) << "RooRealSumFunc::RooRealSumFunc(" << GetName() << ") last func " << func->GetName()
-                               << " is not of type RooAbsReal, fatal error" << endl;
+         auto funcName = func->GetName();
+         coutE(InputArguments) << "RooRealSumFunc::RooRealSumFunc(" << GetName() << ") last func " << (funcName != nullptr ? funcName : "")
+                               << " is not of type RooAbsReal, fatal error\n";
          assert(0);
       }
```
As above.

kgizdov replied:
see comment above

hageboeck commented on roofit/roofitcore/src/RooDataHist.cxx:
```diff
@@ -599,7 +599,8 @@ void RooDataHist::_adjustBinning(RooRealVar &theirVar, const TAxis &axis,
     RooRealVar *ourVar, Int_t *offset)
 {
   if (!dynamic_cast<RooRealVar*>(static_cast<RooAbsArg*>(ourVar))) {
-    coutE(InputArguments) << "RooDataHist::adjustBinning(" << GetName() << ") ERROR: dimension " << ourVar->GetName() << " must be real" << endl ;
+    auto ourVarName = ourVar->GetName();
+    coutE(InputArguments) << "RooDataHist::adjustBinning(" << GetName() << ") ERROR: dimension " << (ourVarName != nullptr ? ourVarName : "undefined") << " must be real\n";
     assert(0) ;
```
Bravo for wanting to improve that error message, but the entire thing is crazy.
- One shouldn't cast `RooRealVar` to `AbsArg` and dynamically back. That's useless.
- If somebody passes something that's not `RooRealVar` or derived into that function, it's time to throw `std::invalid_argument` (maybe better `logic_error`, because it's broken program logic if this happens).
- `assert(0)` is pretty much useless.

@guitargeek do you want to fix this (please?)? If @kgizdov has time to do this, we will happily accept that it, but don't feel obliged.

I suggest something like (github doesn't allow suggestions spanning over deleted lines):
```c++
if ( !dynamic_cast<RooRealVar*>(ourVar) ) throw std::logic_error("Wrong type passed to RooDataHist::_adjustBinning. Please report this issue.");
```

kgizdov replied:
I wasn't sure how much changes you guys were willing to accept. TBH, last night I almost rewrote half the class haha. I think that constructor has broken logic to start with. The comment insists that `numPDF==numCoef or numPDF==numCoef+1`, but it does not actually check for that. It also never validates in the for loop that after it ignores things, the number of PDFs and Coef matches in a desirable way.

There is also a point to be said about the following snippet:
```cpp
// RooAbsCollection.h
  Storage_t::size_type size() const {
    return _list.size();
  }
  ...
  inline Int_t getSize() const { 
    // Return the number of elements in the collection
    return _list.size();
  }
```
and :
```cpp
// RooArgList.
/// Return object at given index, or nullptr if index is out of range
  inline RooAbsArg* at(Int_t idx) const { 

    if (idx >= static_cast<Int_t>(_list.size()))
      return nullptr;

    return _list[idx];
  }
 ...
  /// Access element by index.
  RooAbsArg& operator[](Int_t idx) const {
    assert(0 <= idx && idx < static_cast<Int_t>(_list.size()));
    return *_list[idx];
  }
```
The interface has integer overflows all over.

hageboeck replied:
I know ... 🙁 
There's parts of the interface which have to stay as they are (for backward compatibility), and some that we just didn't get to fix yet. So, for all modernisations that can be done without breaking old code, I'm on board.

For those size issues you mention, I agree that you can accept `std::size_t` as argument. 👍
I'm not sure that you can return `std::size_t` from `getSize()`, though. It's worth a try. The function `size()` is meant as a way out, because it's conforming to what everybody expects, whereas `getSize()` should not be used in new code. If you want to try modernising, I'll be happy to have a look, but @guitargeek has to live with the fallout, so he should say if it's fine to go at it.

As to what you said about `numPDF`:
I'm not sure which class you are talking about, but if you can spare some time, we'll be happy about a helping hand and a PR. If the constructor doesn't check for the number of PDFs, please add those checks. It's better to refuse starting a fit, and then not needing to check while the fit is running.

kgizdov replied:
I was talking about `roofit/roofitcore/src/RooAddModel.cxx`, but I guess in the wrong comment thread hah.

hageboeck reviewed (commented):
Hey Konstantin,

thanks a lot for the contribution! I would ask @guitargeek to validate the logic in the HistFactory parser before approving this.
In the mean time, I have some ideas.

kgizdov commented:
The latest change should address all comments. `clang-format` is applied, and commits are squashed.

@phsft-bot build

jalopezg-r00t commented:
@phsft-bot build

kgizdov commented:
I'm a little confused as the same tests seems to have been quoted as `passed` and `failed` in the same log. Also, I struggle to find the exact definition of each test. I looked through the relevant files, but the structure is not clear to me. Also, last time when I spoke with @amadio about the future of ROOT, he said you guys discussed making sure people outside of CERN can contribute. Unfortunately, I had to use my CERN login to see the test logs. I also don't know how to replicate the tests in my environment. Do they depend on files hosted somewhere? Could you give me some docs to read so I can find my way around?

hageboeck commented:
> I'm a little confused as the same tests seems to have been quoted as `passed` and `failed` in the same log. I struggle to find the exact definition of each test as well. I looked through the relevant files, but the structure is not clear to me.

Which one exactly? I see from the logs that the interpreted stress test exits with code 1, so there's something going on. The node that has the high number of failures is probably a nullptr deref or similar.

> Also, last time when I spoke with @amadio about the future of ROOT, he said you guys discussed making sure people outside of CERN can contribute. Unfortunately, I had to use my CERN login to see the test logs. I also don't know how to replicate the tests in my environment. Do they depend on files hosted somewhere? Could you give me some docs to read so I can find my way around?

Yes, I agree that it's not great to have jenkins behind a CERN login, but I guess this has to be done to secure the instance. The tests should be self-consistent, i.e. bring their own files. Try `ctest -R stressHistFactory -V` to see what's going on locally on your machine (i.e. see invocation and which files are used etc). If the test isn't there, you might have to activate `-Dtesting=ON` or `-Dhistfactory=ON` (which requires `-Dxml=ON`) in cmake.
Does the test pass locally?

hageboeck commented:
I wanted to have a look at where the crash happens, but all commits are now squashed into one. Do you have a version where the commits are separated as the first time when you put up the PR?
Also note that RooFit was merged into ROOT after the projects had been living separately for a while. ROOT uses 3 spaces to indent, but RF has always been using 2. ROOT's clang-format file will therefore introduce a lot of indentation changes, which makes the diffs hard to read.

kgizdov commented:
> Which one exactly? I see from the logs that the interpreted stress test exits with code 1, so there's something going on. The node that has the high number of failures is probably a nullptr deref or similar.

Sorry, I meant that I could not find the test definition in the source. Maybe I'm not looking in the right place. Also, the test in question is quoted two times in the log, once successful and one unsuccessful.

In any case, I do know something is going on and wanted to have a look, but because of the above, I'm not sure where and how.

> Yes, I agree that it's not great to have jenkins behind a CERN login, but I guess this has to be done to secure the instance. The tests should be self-consistent, i.e. bring their own files. Try `ctest -R stressHistFactory -V` to see what's going on locally on your machine (i.e. see invocation and which files are used etc). If the test isn't there, you might have to activate `-Dtesting=ON` or `-Dhistfactory=ON` (which requires `-Dxml=ON`) in cmake. Does the test pass locally?

Alright, I will have a look.

> I wanted to have a look at where the crash happens, but all commits are now squashed into one. Do you have a version where the commits are separated as the first time when you put up the PR? (...)
> ROOT's clang-format file will therefore introduce a lot of indentation changes, which makes the diffs hard to read.

Unfortunately, not. I rebased and force-pushed. But for what it's worth, I think the function in question is rewritten to such an extent that old history before clang-format will not help much. But you can easily use a text editor to switch back to the original indentation. Bear in mind that the original code did not use a consistent indentation - some lines used spaces, others tabs, and the size was different in all places. On top of that, lots of the code was improperly spaced where statements belonging to loops or if's were indented shorter than the scope.

In any case, I will try next week to see what's going on with what you suggested.

hageboeck commented on roofit/histfactory/src/ConfigParser.cxx:
```diff
@@ -260,280 +260,210 @@ std::vector< RooStats::HistFactory::Measurement > ConfigParser::GetMeasurementsF
+   // Then, get the properties of the children nodes
+   TXMLNode *child = node->GetChildren();
+   while (child != nullptr) {
+      const std::string childName(child->GetName()), childNodeName(child->GetNodeName()), childText(child->GetText());
```
```suggestion
      const std::string childName(child->GetName()), childNodeName(child->GetNodeName() ? child->GetNodeName() : ""), childText(child->GetText() ? child->GetText() : "");
```

hageboeck replied:
This fixed the crash for me.

It's admittedly not very easy to find this, since it occurs when `hist2workspace` is invoked, and that happens as a subprocess of the tests that you saw failing.

hageboeck replied:
It's in the logs of the tests, but they don't have line numbers.

One other thing:
> Unfortunately, not. I rebased and force-pushed.

It might be in your `git reflog` if you are lucky.

kgizdov replied:
indeed, constructing a `std::string` from `nullptr` is undefined behaviour... oops. Fixed and pushed. Could you trigger a rebuild?

guitargeek replied:
Sure, thanks for the update!

guitargeek commented:
@phsft-bot build

guitargeek commented:
@phsft-bot build

guitargeek commented:
@phsft-bot build

guitargeek reviewed (approved):
Hi @kgizdov, thanks a lot for the change suggestions! In the meantime RooFit has continued to evolve, and the test failures that we observed in the `testRooWrapperPdf` don't happen anymore. So this PR is good to go for me! If the tests pass I'll merge.

However, since this is not a bugfix PR, I don't think we need to do any backport of this, so I'll close https://github.com/root-project/root/pull/9276. Let me know if that's okay for you!

Thanks
